src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java

```diff
@@ -0,0 +1,181 @@
+                    letterType,
+                    CCDDocument.builder().documentUrl(docAssemblyResponse.getRenditionOutputLocation()).build()
+                ))
+                .build();
+        } catch (DocumentGenerationFailedException e) {
+            logger.info("General Letter creating and preview failed", e);
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .errors(Collections.singletonList(ERROR_MESSAGE))
+                .build();
+        }
+    }
+
+    public CallbackResponse printAndUpdateCaseDocuments(CaseDetails caseDetails, String authorisation) {
+
+        CCDCase ccdCase = caseDetailsConverter.extractCCDCase(caseDetails);
+        CCDDocument draftLetterDoc = ccdCase.getDraftLetterDoc();
+        Claim claim = caseDetailsConverter.extractClaim(caseDetails);
+        List<String> errors = new ArrayList<>();
+        try {
+
+            printLetter(authorisation, draftLetterDoc, claim);
+        } catch (Exception e) {
+            logger.info("General Letter printing and case documents update failed", e);
+            errors = Collections.singletonList(ERROR_MESSAGE);
+        }
+        if (errors.isEmpty()) {
+            logger.info("General Letter: updating case document with general letter");
+            CCDCase updatedCase = ccdCase.toBuilder()
+                .caseDocuments(updateCaseDocumentsWithGeneralLetter(ccdCase, draftLetterDoc))
+                .build();
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .data(caseDetailsConverter.convertToMap(updatedCase))
+                .build();
+        } else {
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .errors(Collections.singletonList(ERROR_MESSAGE))
+                .build();
+        }
+    }
+
+    private List<CCDCollectionElement<CCDClaimDocument>> updateCaseDocumentsWithGeneralLetter(
+        CCDCase ccdCase,
+        CCDDocument draftLetterDoc
+    ) {
+        CCDCollectionElement<CCDClaimDocument> claimDocument = CCDCollectionElement.<CCDClaimDocument>builder()
+            .value(CCDClaimDocument.builder()
+                .documentLink(draftLetterDoc)
+                .documentName(getDocumentNumber(ccdCase))
+                .createdDatetime(LocalDateTime.now(clock.withZone(UTC_ZONE)))
+                .documentType(GENERAL_LETTER)
+                .build())
+            .build();
+        return ImmutableList.<CCDCollectionElement<CCDClaimDocument>>builder()
+            .addAll(ccdCase.getCaseDocuments())
+            .add(claimDocument)
+            .build();
+    }
+
+    private String getDocumentNumber(CCDCase ccdCase) {
+        String number = String.valueOf((ccdCase.getCaseDocuments()
+            .stream()
+            .filter(c -> c.getValue().getDocumentType().equals(GENERAL_LETTER))
+            .filter(c -> c.getValue().getDocumentName().contains(LocalDate.now().toString()))
```

Comment:
Map to the values so you don't need to go through them each time on the filters.
```suggestion
            .stream()
            .map(CCDCollectionElement::getValue)
            .filter(c -> c.getDocumentType().equals(GENERAL_LETTER))
            .filter(c -> c.getDocumentName().contains(LocalDate.now().toString()))
```